### PR TITLE
added dom obj counting to decide sequential/parallel layout (#10110)

### PR DESCRIPTION
**Pull request by avadacatavra**
Added dom count to documents to track number of objects in dom. Changed layout_thread to always create the thread pool, then check to see if it should be parallel/or sequential when it's used.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix part of #10110  (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because @metajack and I aren't sure how to

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11713)

### CODE REVIEW

**Comment by metajack** [comment]
One small omission (-y 1 has changed semantics with this patch and shouldn't) and the rest are nits. r=me with those fixes.

<details><summary><i>Previously, highfive wrote…</i></summary>
> warning **Warning** warning
> 
> * These commits modify layout and script code, but no tests are modified. Please consider adding a test!
</details>


---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[components/layout/layout_thread.rs, line 790 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvCg4_axmyIue7ZYNe:-KJvCg4_axmyIue7ZYNf:179469051) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L790)):*
> ```Rust
>     /// crash.
>     fn exit_now(&mut self) {
>         let ref mut traversal =  self.parallel_traversal;
> ```

Don't add extra space after `=`. One space only.

---

*[components/layout/layout_thread.rs, line 1009 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvAvHUg2pRZpfC6Zn8:-KJvAvHUg2pRZpfC6Zn9:2013444384) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L1009)):*
> ```Rust
>         let document = document.as_document().unwrap();
> 
>         self.parallel_flag = data.dom_count > 600;
> ```

This needs to take the layout threads count into consideration. We don't want a parallel traversal on a thread pool with 1 thread in it, we want the single threaded traversal.

Maybe self.parallel_flag = opts::get().layout_threads > 1 && data.dom_count > 600?

Also, where does 600 come from? @rzambre's suggestion was 750 based on his data. This probably needs a comment documenting where this value came from as well. A link to the bug is sufficient.

---

*[components/layout/layout_thread.rs, line 1136 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvBa8LNzaMqyclEYyH:-KJvBa8LNzaMqyclEYyI:1572330066) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L1136)):*
> ```Rust
>                     || {
>                 // Perform CSS selector matching and flow construction.
>                 match self.parallel_flag {
> ```

Please turn this into a normal if now that's it's a boolean and not an Option.

---

*[components/layout/layout_thread.rs, line 1138 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvBiCagfdhNJrryIws:-KJvBiCagfdhNJrryIwt:-569814214) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L1138)):*
> ```Rust
>                 match self.parallel_flag {
>                     false => {
>                         //Sequential mode
> ```

Space after `//` and `///`. Does tidy not catch that?

---

*[components/layout/layout_thread.rs, line 1144 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvC6YBVe4b4eBGCAw3:-KJvC6YBVe4b4eBGCAw4:976897250) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L1144)):*
> ```Rust
>                     true => {
>                         //Parallel mode
>                         let ref mut traversal = self.parallel_traversal;
> ```

No need for let here. Just pass `&mut self.parallel_traversal` directly below.

---

*[components/layout/layout_thread.rs, line 1403 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvBp5XYYOSB2pR6Th_:-KJvBp5XYYOSB2pR6Tha:-1090244651) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L1403)):*
> ```Rust
> 
>                     debug!("layout: post style parallel? {}", self.parallel_flag);
>                     match self.parallel_flag {
> ```

Normal if please.

---

*[components/layout/layout_thread.rs, line 1410 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvBr0t2GigeNNF9KUR:-KJvBr0t2GigeNNF9KUS:885872438) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L1410)):*
> ```Rust
>                         true => {
>                             // Parallel mode.
>                             let ref mut traversal = self.parallel_traversal;
> ```

As above, skip the let binding and just pass `&mut self.parallel_traversal` directly.

---

*[components/layout/layout_thread.rs, line 1418 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvCRmonZFgMgFfIGb-:-KJvCRmonZFgMgFfIGb0:1611214903) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/layout/layout_thread.rs#L1418)):*
> ```Rust
>                         }
>                     }
> ```

Please don't add new blank lines unless near code you changed.

---

*[components/script/dom/document.rs, line 406 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvDXOjZOpg5A75ntEQ:-KJvDXOjZOpg5A75ntER:-770077250) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/script/dom/document.rs#L406)):*
> ```Rust
> 
>     pub fn increment_dom_count(&self) {
>         self.dom_count.set(self.dom_count.get()+ 1);
> ```

Spaces around `+`.

---

*[components/script/dom/servohtmlparser.rs, line 258 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvDj_u7h3uCcjm42sm:-KJvDj_u7h3uCcjm42sn:-1773544611) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/script/dom/servohtmlparser.rs#L258)):*
> ```Rust
>         };
> ```

Don't add blank lines here. There are no non-whitespace changes in this file, so just revert these two blank line additions.

---

*[components/script/parse/html.rs, line 56 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvDyjC90Y5J5mGtDfg:-KJvDyjC90Y5J5mGtDfh:-1442164879) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/script/parse/html.rs#L56)):*
> ```Rust
> }
> ```

Spurious blank line.

---

*[components/script/parse/html.rs, line 62 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11713#-KJvE-8SgPMZ4dP-Nz4T:-KJvE-8SgPMZ4dP-Nz4U:1700997001) ([raw file](https://github.com/servo/servo/blob/18bc37a162dbd6bc005f1a4d2888d503ee91db77/components/script/parse/html.rs#L62)):*
> ```Rust
> 
>     type Handle = JS<Node>;
> ```

Ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11713#-:-KJvE8fn02XRZc5xp1-T:-1536450404)*

**Comment by dd0x68** [comment]
r @metajack

**Comment by metajack** [comment]
One last nit: You still have no space betwen `//` and `Sequential mode`

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11713#-:-KJvm-UCig7pWzLFA4mu:-1905026090)*

**Comment by metajack** [comment]
@bors-servo r+

This looks great. It will be interesting to see if we see a decrease in test times since most of the wpt tests probably fall into the tiny DOM tree category.

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11713#-:-KJvmpyGkOLpICvk_gQI:-1294747293)*

**Comment by larsbergstrom** [comment]
@bors-servo retry
- #11719

**Comment by metajack** [comment]
@bors-servo p=2

will help in london

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -239,6 +239,8 @@ pub struct Document {
     origin: Origin,
     ///  https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-states
     referrer_policy: Cell<Option<ReferrerPolicy>>,
+    /// track number of dom elements issue #10110
+    dom_count: Cell<i32>,
```
Comment (emilio) [inline components/script/dom/document.rs]: Shouldn't this be an unsigned integer type?

Comment (metajack) [inline components/script/dom/document.rs]: Yes, it probably should. @dd0x68 can you make that change?

**Comment by nox** [comment]
We already have a children counter on `Node`, I wonder if we should add a descendant counter, given it could potentially help some other operations too (for example, a node can't be the child of another if it got more descendants).

**Comment by metajack** [comment]
@bors-servo r+

I don't expect this to pass, as soon of the errors were not intermittents, but maybe we'll get cleaner errors this time.

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11713#-:-KKEHkY680MRJ_JW3VxZ:1036912834)*

**Comment by metajack** [comment]
@bors-servo r+

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11713#-:-KKEhsGIdqv_KQ7XTEmj:1236647940)*

**Comment by dd0x68** [comment]
@bors-servo retry

Testing if intermittent

**Comment by KiChjang** [comment]
What's the status of this? @mbrubeck is on leave now, so who's owning this PR?

**Comment by KiChjang** [comment]
Oh, superseded by #12862.

**Pull request closed without merging**